lib/haskell/natural4/src/LS/Renamer.hs

```diff
@@ -0,0 +1,1295 @@
+  , lsource :: Maybe Text
+  , wwhere :: [RnRule]
+  , srcref :: Maybe SrcRef
+  , defaults :: [RnRelationalPredicate] -- SomeConstant IS 500 ; MentalCapacity TYPICALLY True
+  , symtab :: [RnRelationalPredicate] -- SomeConstant IS 500 ; MentalCapacity TYPICALLY True
+  }
+  deriving (Eq, Ord, Show, Generic)
+
+data RnTypeDecl = RnTypeDecl
+  { name :: RnRuleName -- MyInstance
+  , super :: Maybe RnTypeSig -- IS A Superclass
+  , has :: [RnRule] -- HAS foo :: List Hand, bar :: Optional Restaurant
+  , enums :: Maybe RnParamText -- ONE OF rock, paper, scissors (basically, disjoint subtypes)
+  , given :: Maybe RnParamText
+  , upon :: Maybe RnParamText
+  , rlabel :: Maybe RuleLabel
+  , lsource :: Maybe Text.Text
+  , srcref :: Maybe SrcRef
+  , defaults :: [RnRelationalPredicate] -- SomeConstant IS 500 ; MentalCapacity TYPICALLY True
+  , symtab :: [RnRelationalPredicate] -- SomeConstant IS 500 ; MentalCapacity TYPICALLY True
+  }
+  deriving (Eq, Ord, Show, Generic)
+
+data RnTypeSig
+  = RnSimpleType LS.ParamType RnEntityType
+  | RnInlineEnum LS.ParamType RnParamText
+  deriving (Eq, Ord, Show, Generic)
+
+newtype RnParamText = RnParamText
+  { mkParamText :: NonEmpty RnTypedMulti
+  }
+  deriving (Eq, Ord, Show, Generic)
+
+data RnTypedMulti = RnTypedMulti
+  { rnTypedMultiExpr :: NonEmpty RnExpr
+  , rnTypedMultiTypeSig :: Maybe RnTypeSig
+  }
+  deriving (Eq, Ord, Show, Generic)
+
+-- | A name is something that can be resolved as either a variable, function, or enum.
+data RnName = RnName
+  { rnOccName :: OccName
+  , rnUniqueId :: Unique
+  , rnNameType :: RnNameType
+  -- TODO: add the binding scope for scope checking
+  -- , rnBindingScope :: BindingScope
+  }
+  deriving (Eq, Ord, Show, Generic)
+
+data RnNameType
+  = RnSelector
+  | RnFunction
+  | RnVariable
+  | RnType
+  | RnEnum
+  | RnBuiltin
+  deriving (Eq, Ord, Show, Generic)
+
+data RnExpr
+  = RnExprName RnName
+  | RnExprLit RnLit
+  deriving (Eq, Ord, Show, Generic)
+
+data RnLit
+  = RnInt Integer
+  | RnDouble Double
+  | RnBool Bool
+  | RnString Text
+  deriving (Eq, Ord, Show, Generic)
+
+type RnMultiTerm = [RnExpr]
+
+data RnRelationalPredicate
+  = -- | Might be something like a record access.
+    RnRelationalTerm RnMultiTerm
+  | RnConstraint RnMultiTerm LS.RPRel RnMultiTerm
+  | RnBoolStructR RnMultiTerm LS.RPRel RnBoolStructR
+  | RnNary LS.RPRel [RnRelationalPredicate]
+  deriving (Eq, Ord, Show, Generic)
+
+-- ----------------------------------------------------------------------------
+-- Typed Errors
+-- ----------------------------------------------------------------------------
+
+data RenamerError
+  = UnsupportedRule Text Rule
+  | UnsupportedRPParamText LS.ParamText
+  | UnsupportedUpon LS.ParamText
+  | UnknownMultiTerms LS.MultiTerm
+  | FixArityFunctionNotFound RnName [RnExpr]
+  | ArityErrorLeft !Int RnName [RnExpr]
+  | ArityErrorRight !Int RnName [RnExpr]
+  | UnexpectedNameNotFound OccName
+  | UnexpectedRnNameNotFound RnName
```

Comment:
Minor thing, but as I read this I was wondering what are OccName and RnName, and what's the difference between these two errors. I'm sure I'll find the answers elsewhere, but just reporting that a comment here could be useful.